### PR TITLE
Store upgrade

### PR DESCRIPTION
**Pull request by iprice73**
So here is MVP version of store class with working trading system. Main function is only for testing purposes so it looks awfull. There are still plenty of things that have to be changed. I had to edit some of the existing code in other classes to make this compileable and I'm not sure why some of the things have not worked before like unique pointers. Here is list of things yet to be done:
- unique items per island,
- selling product should increase the amount in the current store (not sure if we want something like this),
- buying items from store should increase their price as amount is decreasing and reverse
- generate functions have some magic values,
- generateItems does not randomize rarity (im not sure how to do it)
- proper << operator in Store and maybe in Ship,
- implement nextDay() in Store

### CODE REVIEW

File: shm/Ship.cpp [hunk 1]
```diff
@@ -41,20 +42,37 @@ Ship& Ship::operator+=(const size_t crew) {
...
+        *cargo -= amount;
+    }
 }
 
-void Ship::unload(Cargo* cargo) {
-    if (!cargo->getAmount()) {
-        std::remove_if(cargo_.begin(), cargo_.end(), [cargo](const auto& ptr) {
-            return ptr.get() == cargo;
-        });
+void Ship::print() const {
```
Comment (J-Cieplinski) [inline shm/Ship.cpp]: I think it could be done with overloaded << operator

File: shm/Player.hpp [hunk 2]
```diff
@@ -8,22 +8,22 @@
 
 class Player : public Delegate {
 private:
-    std::unique_ptr<Ship> ship_;
+    std::shared_ptr<Ship> ship_;
```
Comment (J-Cieplinski) [inline shm/Player.hpp:11]: And why was this changed from unique_ptr?

**Review by J-Cieplinski — CHANGES_REQUESTED** [review]

File: shm/Ship.cpp (hunk 1 above)
Comment (iprice73) [reply shm/Ship.cpp]: yeah but someone else has << operator in store, so I quess he could do this aswell

File: shm/Player.hpp (hunk 2 above)
Comment (iprice73) [reply shm/Player.hpp:11]: it was causing error that object does not exist anymore, im trying to figure it out now

File: shm/Ship.cpp [hunk 3]
```diff
@@ -41,20 +42,37 @@ Ship& Ship::operator+=(const size_t crew) {
 }
 
 void Ship::load(const std::shared_ptr<Cargo>& cargo) {
```
Comment (niemarcin) [inline shm/Ship.cpp]: Take a look at Cargo.hpp file from my branch "crazyShopping" https://github.com/niemarcin/object-oriented-programming/blob/crazyShopping/shm/Cargo.hpp where I proposed two aliases for usage across the whole project. 

I think it might be helpful to define it like that because whenever we decide that we want to change shared_ptr to unique_ptr then we would have less to change in the whole application.

```cpp
using CargoPtr = std::shared_ptr<Cargo>;
using CargoStock = std::vector<std::shared_ptr<Cargo>>;
```

File: shm/Store.hpp [hunk 4]
```diff
@@ -1,9 +1,38 @@
...
+const std::vector<std::string> itemNames = {
+    "cutlass",
+    "chest",
+    "parrot",
+    "old coin",
+    "rusty gun",
+    "pirate's flag"};
+
+constexpr size_t marketSection = 4;
+using Cargos = std::vector<::std::shared_ptr<Cargo>>;
```
Comment (niemarcin) [inline shm/Store.hpp]: As I mentioned earlier, I proposed two aliases in Cargo.hpp file
```cpp
using CargoPtr = std::shared_ptr<Cargo>;
using CargoStock = std::vector<std::shared_ptr<Cargo>>;
```

And I think Cargo.hpp file is better place for it, rather than Store.hpp, because the same CargoStock can be use in Ship class as well as in Store class.

File: shm/Cargo.hpp [hunk 5]
```diff
@@ -17,6 +17,7 @@ class Cargo : public Observer {
     virtual size_t getAmount() const = 0;
     virtual size_t getBasePrice() const = 0;
     virtual size_t getPrice() const = 0;
+    virtual size_t getUniqueStat() const = 0;
```
Comment (niemarcin) [inline shm/Cargo.hpp:22]: I can understand why you create this method and it's nice idea how you later use it in Store::makeCargoToBuy() method in `Alcohol alco(cargo->getName(), amount, cargo->getBasePrice(), cargo->getUniqueStat());` but on the other hand - what if we had a class derived from Cargo where this unique information would not have `size_t` type but for example `double`. Then `getUniqueStat()` would not work.

I thought about something but don't know if it would work. In `Store::makeCargoToBuy` for example in code
```cpp
else if (typeid(*cargo) == typeid(Alcohol)) {
        Alcohol alco(cargo->getName(), amount, cargo->getBasePrice(), cargo->getUniqueStat());
        return std::make_shared<Alcohol>(alco);
```
maybe we can replace `cargo->getUniqueStat()` with `static_cast<Alcohol*>(cargo.get())->getAlcoholContent()`. It's not very nice solution, but then derived cargo can have any type of this special attribute. What do you think?

File: shm/Fruit.hpp [hunk 6]
```diff
@@ -2,13 +2,14 @@
 
 #include "Cargo.hpp"
 
+const size_t expirationDays{10};
```
Comment (niemarcin) [inline shm/Fruit.hpp]: `constexpr` should work here

File: shm/Fruit.hpp [hunk 7]
```diff
@@ -18,15 +19,16 @@ class Fruit : public Cargo {
     size_t getAmount() const override { return amount_; }
     size_t getBasePrice() const override { return basePrice_; }
     size_t getPrice() const override {
-        return static_cast<size_t>(basePrice_ * daysToRot_ / expirationDays_);
+        return static_cast<size_t>(basePrice_ * daysToRot_ / expirationDays);
     }
+    size_t getUniqueStat() const override { return daysToRot_; }
+    size_t getExpirationDays() const { return expirationDays; }
```
Comment (niemarcin) [inline shm/Fruit.hpp]: Why did you move this method to `override from Cargo` section?

File: shm/Fruit.hpp [hunk 8]
```diff
@@ -18,15 +19,16 @@ class Fruit : public Cargo {
...
+
     Cargo& operator+=(const size_t amount) override;
     Cargo& operator-=(const size_t amount) override;
     bool operator==(const Cargo& cargo) const override;
     bool operator!=(const Cargo& cargo) const override;
 
     Fruit& operator--();
 
-    size_t getDaysToRot() const { return daysToRot_; }
-    size_t getExpirationDays() const { return expirationDays_; }
```
Comment (niemarcin) [inline shm/Fruit.hpp:31]: This method should be left here.

File: shm/Player.cpp [hunk 9]
```diff
@@ -2,8 +2,8 @@
 
 #include <numeric>
 
-Player::Player(std::unique_ptr<Ship>& ship, size_t money) {
-    ship_ = std::move(ship);  //it might be good to let the Player class to construct Ship for itself
+Player::Player(std::shared_ptr<Ship>& ship, size_t money) {
+    ship_ = ship; /*std::move(ship);*/  //it might be good to let the Player class to construct Ship for itself
```
Comment (niemarcin) [inline shm/Player.cpp:6]: I think Player should have the Ship as `unique_ptr`.

File: shm/Ship.cpp (hunk 1 above)
Comment (niemarcin) [reply shm/Ship.cpp]: I've done operator<< for Ship class on my branch "crazyShopping", but it was done with `basePrice`

```cpp
std::ostream& operator<<(std::ostream& out, const Ship& ship) {
    std::string horizontalSeparator(61, '=');
    std::string headerSeparator(22, '=');
    int i = 0;
    out << "\n" << headerSeparator << " SHIP'S  STOCK " << headerSeparator
              << "\n"
              << "|| NAME"
              << std::setw(27) << "| QTY "
              << std::setw(5) << "| BASE PRICE "
              << std::setw(3) << " ||\n"
              << horizontalSeparator << "\n";

    for (const CargoPtr& el : ship.cargo_) {
        out << "||"
                  << std::setw(2) << ++i << ". "
                  << std::setw(21) << std::left << el->getName() << " | "
                  << std::setw(3) << std::right << el->getAmount() << " | "
                  << std::setw(10) << std::right << el->getBasePrice() << " ||\n";
    }
    out << horizontalSeparator << "\n";
    
    return out;
}
```

File: shm/Ship.cpp [hunk 10]
```diff
@@ -41,20 +42,37 @@ Ship& Ship::operator+=(const size_t crew) {
...
+    auto it = std::find_if(begin(cargo_), end(cargo_),
+                           [&cargo](const auto& ptr) {
+                               return *cargo == *ptr;
+                           });
+
+    if (it != end(cargo_)) {
+        *(*it) += cargo->getAmount();
+    } else {
+        cargo_.push_back(cargo);
+    }
```
Comment (niemarcin) [inline shm/Ship.cpp]: You can take a look at my branch https://github.com/niemarcin/object-oriented-programming/blob/crazyShopping/shm/Ship.cpp

I have there additional `findGivenCargoInStock()` private method - I think it might be more clear to read it.

But, I have little bit different implementation of `load()` as I also use the `amount` parameter.

File: shm/Store.cpp [hunk 11]
```diff
@@ -1,33 +1,152 @@
...
-    : time_(time) {
-    time_->addObserver(this);
+#include "Alcohol.hpp"
+#include "Fruit.hpp"
+#include "Item.hpp"
+
+Store::Store(std::shared_ptr<Time>& time) {
+    time_ = time;
+    time->addObserver(this);
+    market_ = makeStock(generateFruits(), generateAlcos(), generateItems());
```
Comment (niemarcin) [inline shm/Store.cpp]: I would rather move `generateFruits(), generateAlcos(), generateItems()` into the body of `makeStock()` method. It looks strange when called directly as parameters. Results of these three methods are used only within `makeStock()` so I think it's better to call them inside of `makeStock`.

File: shm/Store.cpp [hunk 12]
```diff
@@ -1,33 +1,152 @@
...
+    result.reserve(marketSection * 3);
+    for (size_t i = 0; i < marketSection; i++) {
+        result.push_back(fruits[i]);
+        result.push_back(alcos[i]);
+        result.push_back(items[i]);
+    }
+    result.shrink_to_fit();
+
+    return result;
+}
```
Comment (niemarcin) [inline shm/Store.cpp]: If you call `generateFruits()`, `generateAlcos()`, etc. inside `makeStock` instead of providing fruits, alcos and items as parameters, then you can have  something like this:
```cpp
void Store::generateCargo() { 
    generateAlcohol();
    generateFruits();
    generateItems();
}
```

and then each `generateSomethin()` method can emplace_back directly to CargoStock of this Store. I think we would have less data copies between functions. 

I've done something like that on my "crazyShopping" branch.

File: shm/Store.cpp [hunk 13]
```diff
@@ -1,33 +1,152 @@
...
+        Item item(itemNames[generateRandom(0, 5)], generateRandom(1, 10), generateRandom(30, 100), Rarity::rare);
+        if(std::none_of(begin(result), end(result),[&item](const auto& ptr){ return ptr->getName() == item.getName();})) {
+            result.emplace_back(std::make_shared<Item>(item));
+            i++;
+        }
+    }
+    result.shrink_to_fit();
+
+    return result;
+}
```
Comment (niemarcin) [inline shm/Store.cpp]: I've seen you comment that you're not sure how to randomize rarity values. 

I've done something like that:
```cpp
void Store::generateItems() {
    for (auto name : itemNames_) {
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> randAmount(0, 20);
        std::uniform_int_distribution<> randBasePrice(50, 100);
        std::uniform_int_distribution<> randRarity(0, rarityAmount - 1);
        cargo_.emplace_back(std::make_shared<Item>(name, randAmount(gen), randBasePrice(gen), possibleRarities[randRarity(gen)]));
    }
}
```

and `rarityAmount` and `possibleRarities` are defined in Item.hpp file as:
```cpp
constexpr size_t rarityAmount{4};
constexpr Rarity possibleRarities[rarityAmount]{Rarity::common, Rarity::rare, Rarity::epic, Rarity::legendary};
```

So it's basically taking a randomized-index element array.

File: shm/Store.cpp [hunk 14]
```diff
@@ -1,33 +1,152 @@
...
+    result.reserve(marketSection);
+
+    size_t i = 0;
+    while(i < marketSection) {
+        Fruit fruit(fruitNames[generateRandom(0, 5)], generateRandom(1, 20), generateRandom(10, 30), generateRandom(1, 10));
+        if(std::none_of(begin(result), end(result),[&fruit](const auto& ptr){ return ptr->getName() == fruit.getName();})) {
+            result.emplace_back(std::make_shared<Fruit>(fruit));
+            i++;
+        }
+    }
```
Comment (niemarcin) [inline shm/Store.cpp]: Is it possible that we would get only one fruit after this code will execute? I mean, if it happens that each time we get the same fruitName as the first one, then we won't add any more fruit to the Stock. Am I right?

**Review by niemarcin — CHANGES_REQUESTED** [review]
That's a nice piece of work!

I didn't post my "crazyShopping" branch yet to PR as it wasn't finished yet, but I can see that we can combine our solutions in one. 

Please take a look at my solution at the mentioned branch. I can create PR so you can comment it line-by-line if you want.

**Comment by niemarcin** [comment]
@LogarithmDerivative Regarding your todo-list, let me comment some of the bullets. 

> * unique items per island,

I'd rather leave it for later.

> * selling product should increase the amount in the current store (not sure if we want something like this),

I think we should implement it.

> * buying items from store should increase their price as amount is decreasing and reverse

I'd rather leave it for later. It's not necessary to implement it and I think we still have much to do.

> * generate functions have some magic values,

Yes indeed, we should get rid of them.

> * generateItems does not randomize rarity (im not sure how to do it)

I've proposed a solution in code review.

> * proper << operator in Store and maybe in Ship,

I've proposed the one for Ship class in code review.

> * implement nextDay() in Store

On my "crazyShopping" branch I've implemented it as
```cpp
void Store::nextDay() {
    cargo_.clear();
    generateCargo();
}
```
Where `generateCargo()` is the same method which is called in constructor for Store object.

File: shm/Ship.cpp (hunk 3 above)
Comment (iprice73) [reply shm/Ship.cpp]: good idea, I will replace it

File: shm/Cargo.hpp (hunk 5 above)
Comment (iprice73) [reply shm/Cargo.hpp:22]: I tried something like this but the problem is that Cargo pointer does not see unique functions from derived classes, `getUniqueStat` is the closest I could get to my idea

File: shm/Fruit.hpp (hunk 7 above)
Comment (iprice73) [reply shm/Fruit.hpp]: oh, I will separete this, just wanted to get all getters in order

File: shm/Player.cpp (hunk 9 above)
Comment (iprice73) [reply shm/Player.cpp:6]: Yes im still working on it, for now unique_ptr throws me some errors but this will be definetly changed

File: shm/Ship.cpp (hunk 10 above)
Comment (iprice73) [reply shm/Ship.cpp]: will replace it, I forgot I left this that ugly

File: shm/Store.cpp (hunk 13 above)
Comment (iprice73) [reply shm/Store.cpp]: I also had an idea to change vector of item names to map<name, rarity>, so we can make unique items and avoid situations like same items having diffrent rarity in various stores. Also in this way we can make legendary items very rare and common items more common

File: shm/Store.cpp (hunk 14 above)
Comment (iprice73) [reply shm/Store.cpp]: We won't add same fruit again but the `i` won't increament aswell so the loop continues untill we get new name

File: shm/Cargo.hpp (hunk 5 above)
Comment (niemarcin) [reply shm/Cargo.hpp:22]: It won't see these functions even after casting to selected class? Like this `static_cast<Alcohol*>(cargo.get())->getAlcoholContent()` ?

File: shm/Store.cpp (hunk 14 above)
Comment (niemarcin) [reply shm/Store.cpp]: Yeah, you're right. My bad.

File: shm/Cargo.hpp (hunk 5 above)
Comment (iprice73) [reply shm/Cargo.hpp:22]: yes it won't :( , `using element_type = class Cargo’ {aka ‘class Cargo’} has no member named getAlco`. I don't think there is a way to make polimorphic poiter see unique functions

File: shm/Ship.cpp [hunk 15]
```diff
@@ -40,21 +42,60 @@ Ship& Ship::operator+=(const size_t crew) {
...
+    if (amount == cargo->getAmount()) {
+        cargo_.erase(std::remove(begin(cargo_), end(cargo_), cargo), cargo_.end());
+    } else {
+        *cargo -= amount;
+    }
+}
+
+std::ostream& operator<<(std::ostream& out, const Ship& ship) {
+    std::string horizontalSeparator(41, '=');
+    std::string headerSeparator(22, '=');
```
Comment (niemarcin) [inline shm/Ship.cpp]: It looks like this `headerSeparator` is not used.

File: shm/Ship.cpp [hunk 16]
```diff
@@ -40,21 +42,60 @@ Ship& Ship::operator+=(const size_t crew) {
...
+    int i = 0;
+    out << "\n" << horizontalSeparator
+              << "\n"
+              << "|| SHIP'S  STOCK"
+              << std::setw(15) << "| QTY "
+              << std::setw(5) << "| PRICE"
+              << std::setw(3) << " ||\n"
+              << horizontalSeparator << "\n";
```
Comment (niemarcin) [inline shm/Ship.cpp]: It might be easier to read the code when the proper `setw` is in the same line with the item it is related to. Similarly to what you have in lines 86-89. What do you think?

File: shm/Ship.hpp [hunk 17]
```diff
@@ -36,14 +36,17 @@ class Ship : public Observer {
     size_t getSpeed() const { return speed_; }
     std::string getName() const { return name_; }
     size_t getId() const { return id_; }
-    Cargo* getCargo(size_t index) const;
-    const std::vector<std::shared_ptr<Cargo>>& getAllCargos() const { return cargo_; }
+    CargoPtr getCargo(size_t index) const;
+    const CargoStock& getAllCargos() const { return cargo_; }
```
Comment (niemarcin) [inline shm/Ship.hpp]: Take a look on line 22 ` std::vector<std::shared_ptr<Cargo>> cargo_;` - you can use the alias there as well.

File: shm/Ship.hpp [hunk 18]
```diff
@@ -36,14 +36,17 @@ class Ship : public Observer {
...
 
     Ship& operator-=(const size_t crew);
     Ship& operator+=(const size_t crew);
 
-    void load(const std::shared_ptr<Cargo>& cargo);
-    void unload(Cargo* cargo);
+    CargoPtr findCargo(const CargoPtr& cargo) const;    
+    void load(const CargoPtr& cargo);
+    void unload(const CargoPtr& cargo, size_t amount);
```
Comment (niemarcin) [inline shm/Ship.hpp:46]: Just asking - the `cargo` parameter type for `unload` method was changed from Cargo* to CargoPtr which is an alias for `std::shared_ptr<Cargo>` - was this change from Cargo* to shared_ptr done intentionally or by accident? The same change was done in getCargo method.

File: shm/Ship.cpp [hunk 19]
```diff
@@ -40,21 +42,60 @@ Ship& Ship::operator+=(const size_t crew) {
...
-void Ship::load(const std::shared_ptr<Cargo>& cargo) {
-    cargo_.push_back(cargo);
+CargoPtr Ship::findCargo(const CargoPtr& cargo) const {
+    auto it = std::find_if(begin(cargo_), end(cargo_),
+                           [&cargo](const auto& ptr) {
+                               return *cargo == *ptr;
+                           });
+    
+    return (it == end(cargo_)) ? nullptr : *it;
 }
```
Comment (niemarcin) [inline shm/Ship.cpp]: It seems like we have almost the same method findCargo in Ship.cpp and in Store.cpp but the container differs (cargo_ or market_ ). So it's kind of duplicated code. Maybe we can extract findCargo as a static method of Cargo class with given two parametes - cargoToFind and stockToSearch (or something like this)?

File: shm/Player.hpp [hunk 20]
```diff
@@ -8,22 +8,22 @@
...
 
 public:
-    Player(std::unique_ptr<Ship>& ship, size_t money);
+    Player(std::shared_ptr<Ship>& ship, size_t money);
 
     size_t getMoney() const { return money_; };
     size_t getAvailableSpace() const { return availableSpace_; };
     size_t getSpeed() const { return ship_->getSpeed(); };
-    Cargo* getCargo(size_t index) const { return ship_->getCargo(index); };
+    std::shared_ptr<Cargo> getCargo(size_t index) const { return ship_->getCargo(index); };
```
Comment (niemarcin) [inline shm/Player.hpp]: Use alias for CargoPtr

File: shm/Player.hpp [hunk 21]
```diff
@@ -8,22 +8,22 @@
...
     
-    void purchaseCargo(std::shared_ptr<Cargo> cargo, size_t amount, size_t price);
-    void sellCargo(std::shared_ptr<Cargo> cargo, size_t amount, size_t price);
+    void purchaseCargo(const std::shared_ptr<Cargo>& cargo, size_t amount, size_t price);
+    void sellCargo(const std::shared_ptr<Cargo>& cargo, size_t amount, size_t price);
```
Comment (niemarcin) [inline shm/Player.hpp]: Use alias for CargoPtr

File: shm/Store.cpp [hunk 22]
```diff
@@ -1,33 +1,170 @@
...
+    time->addObserver(this);
+    makeStock();
+}
+
+size_t Store::genRand(int min, int max) const {
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    std::uniform_int_distribution<> distirb(min, max);
+
+    return static_cast<size_t>(distirb(gen));
```
Comment (niemarcin) [inline shm/Store.cpp:25]: Isn't it better to have `min` and `max` parameter as a `size_t` type?

File: shm/Store.cpp [hunk 23]
```diff
@@ -1,33 +1,170 @@
...
+        Item item(itemNames[genRand(0, 5)], genRand(1, 10), genRand(30, 100), possibleRarities[genRand(0, 3)]);
+        if(std::none_of(begin(market_), end(market_),[&item](const auto& ptr){ return ptr->getName() == item.getName();})) {
+            market_.emplace_back(std::make_shared<Item>(item));
+            i++;
+        }
+    }
+}
+
+void Store::makeStock() {
+    market_.reserve(marketSection * 3);
```
Comment (niemarcin) [inline shm/Store.cpp]: Magic number - maybe call it `numberOfCargoTypes` or something similar?

File: shm/Store.cpp [hunk 24]
```diff
@@ -1,33 +1,170 @@
...
+    generateFruits();
+    generateAlcos();
+    generateItems();
+    market_.shrink_to_fit();
```
Comment (niemarcin) [inline shm/Store.cpp:67]: Is it necessary to shrink it?

File: shm/Store.cpp [hunk 25]
```diff
@@ -1,33 +1,170 @@
...
 
-Response Store::buy(Cargo* cargo, size_t amount, Player* player) {
-    const size_t price = amount * cargo->getBasePrice();
+void Store::removeFromStore(const CargoPtr& cargo, size_t amount) {
+    if (amount == cargo->getAmount()) {
+        market_.erase(std::remove(begin(market_), end(market_), cargo), market_.end());
+    } else {
+        *cargo -= amount;
+    }
+}
```
Comment (niemarcin) [inline shm/Store.cpp]: Pretty much the same code as `unload()` in Ship class. Maybe we can refactor it? Similarly to what I wrote about `findCargo` method.

File: shm/Store.cpp [hunk 26]
```diff
@@ -1,33 +1,170 @@
...
+
+void Store::addToStore(const CargoPtr& cargo, size_t amount) {
+    auto ptr = findCargo(cargo);
+
+    if(ptr) {
+        *ptr += amount;
+    } else {
+        market_.push_back(cargo);
+    }
+}
```
Comment (niemarcin) [inline shm/Store.cpp]: Pretty much the same as Ship::load() method. Refactor?

**Review by niemarcin — CHANGES_REQUESTED** [review]

File: shm/Store.hpp [hunk 27]
```diff
@@ -1,9 +1,39 @@
 #pragma once
 
+#include <iostream>
+#include <map>
```
Comment (niemarcin) [inline shm/Store.hpp]: If I'm correct then we don't need `map` here.

File: shm/Ship.cpp (hunk 16 above)
Comment (iprice73) [reply shm/Ship.cpp]: But these are not related to any items, this is just header for our stock

File: shm/Ship.hpp (hunk 18 above)
Comment (iprice73) [reply shm/Ship.hpp:46]: Yeah I changed it because of aliases to make it look more clear, we dont really need raw pointers in here

File: shm/Ship.cpp (hunk 19 above)
Comment (iprice73) [reply shm/Ship.cpp]: oh, good idea!

File: shm/Store.cpp (hunk 24 above)
Comment (iprice73) [reply shm/Store.cpp:67]: On one of the lessons there was metioned that we use this when we reserve memory, it doesn't change anything but I left it just in case

File: shm/Ship.cpp (hunk 16 above)
Comment (niemarcin) [reply shm/Ship.cpp]: I probably wrote it wrong - what I meant was to put `setw` in the same line with the column name it is related to. For example `setw(15)` in the same line as "|| SHIP'S STOCK" and `setw(5)` in the same line as "| QTY " and so on. What do you think? Something like this:

```
    out << "\n" << horizontalSeparator
              << "\n"
              << "|| SHIP'S  STOCK" << std::setw(15)
              << "| QTY "<< std::setw(5)
               << "| PRICE"<< std::setw(3) 
              << " ||\n"
              << horizontalSeparator << "\n";
```

Or maybe I interpret this `setw` incorrectly and it's not related to columns' width?

Comment (iprice73) [reply shm/Ship.cpp]: seems good to me, I will change this both here and Store

File: shm/StockManagement.cpp [hunk 28]
```diff
@@ -0,0 +1,29 @@
+#include "StockManagement.hpp"
+#include <algorithm>
```
Comment (niemarcin) [inline shm/StockManagement.cpp:3]: Add new line between these includes

File: shm/StockManagement.cpp [hunk 29]
```diff
@@ -0,0 +1,29 @@
...
+
+CargoPtr StockManagement::findCargo(const CargoStock& stock_, const CargoPtr& cargo) {
+    auto it = std::find_if(begin(stock_), end(stock_),
+    [&cargo](const auto& ptr) {
+        return  *cargo == *ptr;
+    });
```
Comment (niemarcin) [inline shm/StockManagement.cpp]: Correct formatting - lines 6-7 could have one more level of indentation.

File: shm/StockManagement.cpp [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+        stock_.erase(std::remove(begin(stock_), end(stock_), cargo), stock_.end());
+    } else {
+        *cargo -= amount;
+    }
+}
+
+void StockManagement::addCargoToStock(CargoStock& stock_, const CargoPtr& cargo, size_t amount) {
+    auto ptr = findCargo(stock_, cargo);
+
+    if(ptr) {
```
Comment (niemarcin) [inline shm/StockManagement.cpp]: formatting - `if (ptr)` - spacing

File: shm/Store.hpp [hunk 31]
```diff
@@ -11,15 +41,26 @@ enum class Response {
     lack_of_space
 };
 
-class Store : public Observer {
+class Store : public Observer , public StockManagement {
```
Comment (niemarcin) [inline shm/Store.hpp]: We don't need inheritance here as in StockManagement we only have static methods.

Also, if we remove inheritance then `#include "StockManagement.hpp"` can be probably moved to Store.cpp file.

File: shm/Store.hpp [hunk 32]
```diff
@@ -1,8 +1,38 @@
...
+
+const std::vector<std::string> itemNames = {
+    "cutlass",
+    "chest",
+    "parrot",
+    "old coin",
+    "rusty gun",
+    "pirate's flag"};  
+
+constexpr size_t marketSection = 4;
```
Comment (niemarcin) [inline shm/Store.hpp]: `constexpr size_t marketSection{4};` this initialization style will be more consistent with the rest of our code

File: shm/Store.cpp [hunk 33]
```diff
@@ -1,33 +1,145 @@
...
         return Response::lack_of_space;
     }
-    *cargo += amount;
-    player->purchaseCargo(std::shared_ptr<Cargo> (cargo), amount, price);
+
+    auto cargoToBuy = makeNewCargo(cargo, amount);
+    player->purchaseCargo(cargoToBuy, amount, price);
+    removeCargoFromStock(market_, cargo, amount);
+    
+
```
Comment (niemarcin) [inline shm/Store.cpp:99]: not necessary new line

File: shm/Store.cpp [hunk 34]
```diff
@@ -1,33 +1,145 @@
...
         return Response::lack_of_money;
     } else if (player->getAvailableSpace() < amount) {
         return Response::lack_of_space;
     }
-    *cargo += amount;
-    player->purchaseCargo(std::shared_ptr<Cargo> (cargo), amount, price);
+
+    auto cargoToBuy = makeNewCargo(cargo, amount);
+    player->purchaseCargo(cargoToBuy, amount, price);
+    removeCargoFromStock(market_, cargo, amount);
```
Comment (niemarcin) [inline shm/Store.cpp]: This works because Store class inherits from StockManagement class, but if we decide to remove inheritance then here it should be `StockManagement::removeCargoFromStock( ... ) ;`

File: shm/Store.cpp [hunk 35]
```diff
@@ -1,33 +1,145 @@
...
+        return std::make_shared<Alcohol>(alco);
+    } else if (typeid(*cargo) == typeid(Item)) {
+        Item item(cargo->getName(), amount, cargo->getBasePrice(), static_cast<Rarity>(cargo->getUniqueStat()));
+        return std::make_shared<Item>(item);
+    }
+
+    return nullptr;
+}
+
+CargoPtr Store::getCargo(size_t index) const {
```
Comment (niemarcin) [inline shm/Store.cpp]: Is it possible that provided index will be equal to zero? If so then what will happen in `market_[index-1];` ?

File: shm/Ship.hpp [hunk 36]
```diff
@@ -7,21 +7,22 @@
 #include "Cargo.hpp"
 #include "Delegate.hpp"
 #include "Observer.hpp"
+#include "StockManagement.hpp"
 #include "Time.hpp"
 
 constexpr size_t salaryPerWorker = 1;
 
-class Ship : public Observer {
+class Ship : public Observer, public StockManagement {
```
Comment (niemarcin) [inline shm/Ship.hpp]: We don't need inheritance here as in StockManagement we only have static methods.

Also, if we remove inheritance then `#include "StockManagement.hpp"` can be probably moved to Ship.cpp file.

File: shm/Store.cpp [hunk 37]
```diff
@@ -1,33 +1,145 @@
...
-Response Store::sell(Cargo* cargo, size_t amount, Player* player) {
-    const size_t price = amount * cargo->getPrice();
-    *cargo -= amount; 
-    player->sellCargo(std::shared_ptr<Cargo> (cargo), amount, price);
+Response Store::sell(const CargoPtr& cargo, size_t amount, const std::shared_ptr<Player>& player) {
+    size_t price = cargo->getPrice() * amount;
+
+    auto cargoToSell = makeNewCargo(cargo, amount);
+    player->sellCargo(cargo, amount, price);
+    addCargoToStock(market_, cargoToSell, amount);
```
Comment (niemarcin) [inline shm/Store.cpp]: This works because Store class inherits from StockManagement class, but if we decide to remove inheritance then here it should be `StockManagement::addCargoToStock( ... ) ;`

File: shm/Ship.cpp [hunk 38]
```diff
@@ -15,11 +17,11 @@ Ship::Ship(int maxCrew, int speed, size_t id, std::shared_ptr<Time>& time)
     : Ship(0, maxCrew, speed, "", id, time) {
 }
 
-Cargo* Ship::getCargo(size_t index) const {
-    if (index > cargo_.size() || index == 0) {
+CargoPtr Ship::getCargo(size_t index) const {
```
Comment (niemarcin) [inline shm/Ship.cpp]: It looks like `getCargo()` method from Ship class and Store are pretty similar and they could be also moved to StockManagement class. What do you think? Of course it requires to add one more parameter to list of arguments.

**Review by niemarcin — CHANGES_REQUESTED** [review]

File: shm/Store.cpp (hunk 35 above)
Comment (iprice73) [reply shm/Store.cpp]: This will be validated in my other issue about selling in game class, so this function will be called only after correct input

File: shm/StockManagement.hpp [hunk 39]
```diff
@@ -0,0 +1,13 @@
...
+
+#include "Cargo.hpp"
+
+class StockManagement {
+private:
+    static CargoPtr findCargo(const CargoStock& stock_, const CargoPtr& cargo);
+
+public:
+    static void removeCargoFromStock(CargoStock& stock_, const CargoPtr& cargo, size_t amount);
+    static void addCargoToStock(CargoStock& stock_, const CargoPtr& cargo, size_t amount);
```
Comment (niemarcin) [inline shm/StockManagement.hpp]: Probably it is a mistake from my temporary branch, but please change `stock_` name to `stock`. The same in StockManagement.cpp file.

**Review by niemarcin — CHANGES_REQUESTED** [review]

**Review by niemarcin — APPROVED** [review]
